dipy/denoise/randommatrix_localpca_parallel.pyx

```diff
@@ -0,0 +1,301 @@
+
+
+    sizes[0] = arr.shape[0]
+    sizes[1] = arr.shape[1]
+    sizes[2] = arr.shape[2]
+        
+    noise_arr = np.zeros([arr.shape[0],arr.shape[1],arr.shape[2]],dtype=np.float64)
+    denoised_arr = np.zeros(arr.shape, dtype=np.float64) 
+
+    cdef double [:,:,:,:] denoised_arr_view =denoised_arr
+    cdef double [:,:,:] noise_arr_view =noise_arr
+
+    # OPENMP loop over slices
+    for k in prange(patch_radius,sizes[2]-patch_radius,nogil=True,schedule=static):
+        for j in range(patch_radius,sizes[1]-patch_radius):
+            for i in range(patch_radius,sizes[0]-patch_radius):
+             
+                # copy the local patch into array X
+
+                cnt=-1
+                for ii in range(i-patch_radius,i+patch_radius+1):
+                    for jj in range(j-patch_radius,j+patch_radius+1):
+                         for kk in range(k-patch_radius,k+patch_radius+1):
+                            cnt=cnt+1                                                           
+                            for vol in range(mm):
+                                with cython.boundscheck(False):
+                                    X[k,cnt,vol]=arr_view[ii,jj,kk,vol]
+             
+                # X^TX or XX^T
+                fast_dgemm(X[k,:, :], C[k,:, :])
+                fast_eig(C[k,:, :], W[k,:], WORK[k,:], LWORK,IWORK[k,:],LIWORK) 
+
+
+                # noise eigenvalue cutoff computations
+                lam_r = W[k,0] / nn
+                clam = 0.
+                sigma2 = 0.
+                cutoff_p = 0
+             
+                for p in range(rr):
+                    lam = W[k,p] /nn
+                    clam = clam + lam
+                    gam =  float(mm -rr + p +1) / float(nn)
+                    sigsq1 = clam / (p +1) / max(gam,1.)
+                    sigsq2 = (lam - lam_r) / 4. / sqrt(gam)
+
+                    if sigsq2 < sigsq1:
+                        sigma2= sigsq1
+                        cutoff_p = p +1               
+
+                noise_arr_view[i,j,k]=sqrt(sigma2)               
+ 
+
+                # Reconstruct the donoised image
+                for p in range(cutoff_p):
+                    diag_eigmat[k,p,p]=0
+                for p in range(cutoff_p,rr):
+                    diag_eigmat[k,p,p]=1
+
+                if mm<=nn:
+                    fast_matvec('t',C[k,:,:], X[k,nn/2,:], temp2[k,:])    
+                    fast_matvec('n',diag_eigmat[k,:,:],temp2[k,:],temp3[k,:])
+                    fast_matvec('n',C[k,:,:],temp3[k,:],temp2[k,:])
+              
+                else:
+                    for ii in range(rr):
+                        temp1[k,ii]= C[k,ii,nn/2]
+
+                    fast_matvec('n',diag_eigmat[k,:,:], temp1[k,:] ,  temp0[k,:])
+                    fast_matvec('n',C[k,:,:], temp0[k,:],   temp1[k,:])
+                    fast_matvec('n',X[k,:,:], temp1[k,:],   temp2[k,:])
+                
+                denoised_arr_view[i,j,k,:] = temp2[k,:]
```

Comment:
You are only using the center voxel of each patch for denoising. Please consider using all voxels like using an overcomplete average approach (Manjon et al., 2013)[https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0073021]. In this way, we avoid removing information in cases that tissues are near to the limits of the image or in cases that patch size is large. Keep in mind that some researcher might use PCA in all voxels of an image simultaneously